base/broadcast.jl

```diff
@@ -186,33 +189,87 @@ BroadcastStyle(::MatrixStyle, ::DefaultArrayStyle{N}) where N = DefaultArrayStyl
 BroadcastStyle(::VectorStyle, ::MatrixStyle) = MatrixStyle()
 # end FIXME
 
+### Lazy-wrapper for broadcasting
+
+# `Broadcasted` wrap the arguments to `broadcast(f, args...)`. A statement like
+#    y = x .* (x .+ 1)
+# will result in code that is essentially
+#    y = copy(Broadcasted(*, x, Broadcasted(+, x, 1)))
+# `broadcast!` results in `copyto!(dest, Broadcasted(...))`.
+
+# Besides the function `f` and the input `args`, `Broadcasted`
+# includes two other fields (`axes` and `indexing`) that, once
+# initialized, improve performance when extracting values.  However,
+# in some cases (e.g., StaticArrays.jl) these are not used, and for
+# performance it's important to be able to bypass their
+# initialization. We use `Nothing` type parameters when these have not
+# been intialized.
+
+# The use of `Nothing` in place of a `BroadcastStyle` has a different
+# application, in the fallback method
+#    copyto!(dest, bc::Broadcasted) = copyto!(dest, convert(Broadcasted{Nothing}, bc))
+# This allows methods
+#    copyto!(dest::DestType,  bc::Broadcasted{Nothing})
+# that specialize on `DestType` to be easily disambiguated from
+# methods that instead specialize on `BroadcastStyle`,
+#    copyto!(dest::AbstractArray, bc::Broadcasted{MyStyle})
+
+struct Broadcasted{Style<:Union{Nothing,BroadcastStyle}, ElType, Axes, Indexing<:Union{Nothing,TupleLL}, F, Args<:TupleLL}
+    f::F
+    args::Args
+    axes::Axes          # the axes of the resulting object (may be bigger than implied by `args` if this is nested inside a larger `Broadcasted`)
+    indexing::Indexing  # index-replacement info computed from `newindexer` below
```

Comment:
Does this property really need to be a `TupleLL`?  Could a normal `Tuple` work here?  I'm still working my way through here, so I may be missing something.

Reply:
IIRC a normal tuple would be fine. It just seemed easier to sync with respect to `args` this way. Also since lispy-programming with `TupleLL`s is O(N) on the compiler whereas `Tuple`s are O(N^2), I didn't think it was a bad thing.

Reply:
fwiw, those efficiency numbers aren't quite right due to various compiler optimizations and the violations of the simplistic pointer-chasing model provided by lisp. Either could usually range from `O(N)` to `O(N^2)`, depending on the various codepaths hit. I could have a type that just specialized on the leading argument (the only place where it usually mattered), but figured it would be more sensible to make it symmetric.